File: backend/api/category/file_utils.py
import os
import json
from django.utils.text import slugify


def generate_category_file(category):
    """Tạo file JSON với thông tin phân cấp Category, Task và Command."""
    if category.parent is None:  # Chỉ thực hiện với category gốc
        filename = f"{slugify(category.name)}.json"
        filepath = os.path.join(
            r'C:\Users\phamd\OneDrive\Đồ án SAA\SAA\agent\command',
            filename
        )
        data = _generate_category_hierarchy(category)
        with open(filepath, 'w', encoding='utf-8') as file:
            json.dump(data, file, indent=4, ensure_ascii=False)
        print(f"JSON file created: {filepath}")


def _generate_category_hierarchy(category):
    """Tạo cấu trúc phân cấp của Category, Task và Command."""
    # Dữ liệu cơ bản cho Category
    category_data = {
        "category": category.name,
        "slug": category.slug,
        "command": category.cmd or "No command",
        "expected_output": category.expect_output or "No expected output",
        "tasks": [],
        "children": []  # Danh sách Category con
    }

    # Lặp qua các Task của Category và thêm vào danh sách
    for task in category.tasks.all():
        task_data = {
            "task": task.name,
            "slug": task.slug,
            "combine": task.combine,
            "scored": task.scored,
            "remediation": task.remediation or "No remediation",
            "note": task.note or "No note",
            "commands": []
        }

        # Lặp qua các Command trong Task
        for command in task.commands.all():
            task_data["commands"].append({
                "command": command.cmd,
                "expected_output": command.expect or "No expected output",
                "operator": command.operator or "No operator",
                "parser": command.parser or "No parser"
            })

        # Thêm Task vào danh sách của Category
        category_data["tasks"].append(task_data)

    # Đệ quy xử lý các Category con
    for child_category in category.children.all():
        category_data["children"].append(_generate_category_hierarchy(child_category))

    return category_data


def delete_category_file(category):
    """
    Xóa file JSON của một category nếu tồn tại.
    """
    if category.parent is None:  # Chỉ xóa file của category cha
        filename = f"{slugify(category.name)}.json"
        filepath = os.path.join(
            r'C:\Users\phamd\OneDrive\Đồ án SAA\SAA\agent\command',
            filename
        )
        if os.path.exists(filepath):
            os.remove(filepath)
            print(f"Deleted file: {filepath}")
        else:
            print(f"File does not exist: {filepath}")

File: backend_nodejs/utils/fileUtils.js
const fs = require('fs');
const path = require('path');
const slugify = require('slugify');

// Tạo file JSON cho Category gốc
exports.generateCategoryFile = async (category) => {
    try {
        if (!category.parentId) { // Chỉ xử lý với category gốc
            const filename = `${slugify(category.name, { lower: true })}.json`;
            const filepath = path.join(__dirname, '../../agent/command', filename);

            const data = await this.generateCategoryHierarchy(category);
            fs.writeFileSync(filepath, JSON.stringify(data, null, 4), 'utf-8');
            console.log(`JSON file created: ${filepath}`);
        }
    } catch (error) {
        console.error('Error generating category file:', error.message);
    }
};

// Tạo cấu trúc phân cấp Category, Task, Command
exports.generateCategoryHierarchy = async (category) => {
    try {
        const tasks = await category.getTasks();
        const children = await category.getChildren();

        return {
            category: category.name,
            slug: category.slug,
            command: category.cmd || 'No command',
            expected_output: category.expect_output || 'No expected output',
            tasks: await Promise.all(tasks.map(async (task) => ({
                task: task.name,
                slug: task.slug,
                combine: task.combine,
                scored: task.scored,
                remediation: task.remediation || 'No remediation',
                note: task.note || 'No note',
                commands: await task.getCommands().then((commands) =>
                    commands.map((cmd) => ({
                        operator: cmd.operator,
                        cmd: cmd.cmd,
                        expect: cmd.expect,
                        parser: cmd.parser || 'No parser',
                    }))
                ),
            }))),
            children: await Promise.all(children.map(this.generateCategoryHierarchy)),
        };
    } catch (error) {
        console.error('Error generating category hierarchy:', error.message);
        return {};
    }
};

// Xóa file JSON của category
exports.deleteCategoryFile = (category) => {
    try {
        if (!category.parentId) {
            const filename = `${slugify(category.name, { lower: true })}.json`;
            const filepath = path.join(__dirname, '../../agent/command', filename);
            if (fs.existsSync(filepath)) {
                fs.unlinkSync(filepath);
                console.log(`Deleted file: ${filepath}`);
            } else {
                console.warn(`File not found: ${filepath}`);
            }
        }
    } catch (error) {
        console.error('Error deleting category file:', error.message);
    }
};
